frontend/src/Component/Pages/Homepage.jsx

```diff
@@ -17,44 +18,70 @@ export const Homepage = () => {
-          dev_eui: device.dev_eui,
+          dev_eui: device.dev_eui
         }));
         setDevices(extractedData);
 
         const co2Data = {};
-        await Promise.all(extractedData.map(async (device) => {
-          // This fetches the co2 level for each room
-          const co2Response = await fetch(`${apiKey}/api/v1/rooms/latest/${device.dev_eui}`);
-          const co2Info = await co2Response.json();
-          co2Data[device.dev_eui] = co2Info.data.co2;
-        }));
+        await Promise.all(
+          extractedData.map(async (device) => {
+            // This fetches the co2 level for each room
+            const co2Response = await fetch(
+              `${apiKey}/api/v1/rooms/latest/${device.dev_eui}`
+            );
+            const co2Info = await co2Response.json();
+            co2Data[device.dev_eui] = co2Info.data.co2;
+          })
+        );
         setCo2Levels(co2Data);
         setTemperatures(tempData);
+        setIsLoading(false);
       } catch (error) {
         console.error('Error fetching devices or CO2 levels:', error);
+        setIsLoading(false);
       }
     };
     fetchDevices();
   }, [socket]);
-  
+
   return (
     <div className="pt-20 text-center">
-      <h1 data-cy="h1Welcome" className="text-6xl ">Welcome to D-Block CO<sub>2</sub> Monitor</h1>
-
-      <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4">
-        {/* Create one grid container outside the loop */}
-        {devices.map((device) => (
-          <div key={device.dev_eui} className="flex justify-center">
-            <li>
-              <NavLink to={`/D-Block/${device.room_number}`} className="link">
-                <Co2Sensor room_number={device.room_number} co2={co2Levels[device.dev_eui] || 400} size="max-content"/>
-              </NavLink>
-            </li>
-          </div>
-        ))}
-      </div>
+      <h1 data-cy="h1Welcome" className="text-6xl ">
+        Welcome to D-Block CO<sub>2</sub> Monitor
+      </h1>
+      <>
+        {isLoading ? (
+          <>
+            <LoadingSpinner />
+          </>
+        ) : devices ? (
+          <>
+          {/* Create one grid container outside the loop */}
+            <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4">
+              {devices.map((device) => (
+                <div key={device.dev_eui} className="flex justify-center">
+                  <li>
+                    <NavLink
+                      to={`/D-Block/${device.room_number}`}
+                      className="link"
+                    >
+                      <Co2Sensor
+                        room_number={device.room_number}
+                        co2={co2Levels[device.dev_eui] || 400}
+                        size="max-content"
+                      />
+                    </NavLink>
+                  </li>
+                </div>
+              ))}
+            </div>
+          </>
+        ) : (
+          <p>Failed to load data</p>
```

Comment:
As with the error state above, we can change this Failed to load data, to a error && <p>{error}</p> instead. Something we can talk about throughout the week.